### PR TITLE
Ensure IdP certificate is formatted properly

### PR DESCRIPTION
**Pull request by phlipper**
When loading the provider certificate from the environment, or integrating via `omniauth-saml`, we would often encouter a series of _nested asn1 error_ messages due to the malformed certificate failing to be read by `OpenSSL`.

Inspiration for this patch comes from toyokazu/rack-saml@b7646ad

Thanks to @dannyb for the assistance!

### CODE REVIEW

**Comment by buffym** [comment]
:+1:

**Comment by Lordnibbler** [comment]
:+1: this looks nice, thanks for the improvement.  @pitbulk @luisvm can you review please?

**Comment by luisvm** [comment]
:+1: lgtm

**Comment by Lordnibbler** [comment]
@pitbulk can you review this? seems good :+1:

File: lib/onelogin/ruby-saml/utils.rb [hunk 1]
```diff
@@ -1,19 +1,16 @@
...
+        # don't try to format an encoded certificate
+        return cert if cert.match(/\x0d/)
 
-          if heads
-            cert = cert.scan(/.{1,64}/).join("\n")+"\n"
-            cert = "-----BEGIN CERTIFICATE-----\n" + cert + "-----END CERTIFICATE-----\n"
-          end
-        end
-        cert
+        cert = cert.gsub(/\-{5}\s?(BEGIN|END) CERTIFICATE\s?\-{5}/, "")
```
Comment (Lordnibbler) [inline lib/onelogin/ruby-saml/utils.rb]: any chance its possible to do this certificate parsing using the `OpenSSL` module instead of `gsub` and `scan`?

**Comment by pitbulk** [comment]
I integrated some partof this PR at the response_validation branch:
https://github.com/onelogin/ruby-saml/commit/da8c9ec1679cbe66ec6be5411b48a9faa63f9a28
So we can discard that PR.

OpenSSL can't help to format the cert in the way that we did.
We don't want that the toolkit fail due a cert or private key contain an extra space due copy/paste :)

File: lib/onelogin/ruby-saml/utils.rb (hunk 1 above)
Comment (phlipper) [inline lib/onelogin/ruby-saml/utils.rb]: @Lordnibbler I doubt it. I don't see anything in the [OpenSSL module docs](http://ruby-doc.org/stdlib-2.2.1/libdoc/openssl/rdoc/OpenSSL.html) that handles invalid certificate formatting. Trying to load an invalid certificate with the `OpenSSL` module results in the `nested asn1 error`, which is the point of this PR.

```
irb(main):001:0> OpenSSL::X509::Certificate.new File.read("./test/certificates/invalid_certificate1")
OpenSSL::X509::CertificateError: nested asn1 error
```

Comment (Lordnibbler) [inline lib/onelogin/ruby-saml/utils.rb]: thanks for looking, just thought I'd ask :smile:

**Comment by phlipper** [comment]
Is there anything else I can do to help get this merged? :)

**Comment by Lordnibbler** [comment]
@pitbulk so just to clarify, do you NOT want to merge this branch? seems like its ready to go, and you could incorporate these changes simply by rebasing your response_validation branch off of master?

**Comment by pitbulk** [comment]
@Lordnibbler  I already added the "Ensure IdP certificate" fix in  #197, and I will refactor the tests once the 197 PR and the encryption PR merged, otherwise gonna be a chaos.

**Comment by Lordnibbler** [comment]
@pitbulk wouldn't it be better to get this tiny ~100 LOC change in, rather than try to merge it as part of #197 which is giant ~3000 LOC change?

**Comment by Lordnibbler** [comment]
@pitbulk can we merge this?

**Comment by pitbulk** [comment]
If we don't gonna merge the 197 PR, you can merge it.

**Comment by phlipper** [comment]
Is there any chance we can get this in before 0.9.2 is cut?

**Comment by phlipper** [comment]
Can someone please explain what the issue/holdup is with this PR? It is almost 5 months old...

We have several others that we'd like submit back upstream, but it's unclear what the expectations and timelines are. Thanks.

**Comment by pitbulk** [comment]
@phlipper I'm splitting the #197 PR and plan to add this PR too.

**Comment by pitbulk** [comment]
@phlipper I rebased this PR and plan to merge it. #226

Sorry for the delay

**Comment by pitbulk** [comment]
@phlipper 

I was adding test to the setting view and when I tried to execute:

```
        @settings = OneLogin::RubySaml::Settings.new
        @settings.certificate = read_certificate("valid_certificate")
        @settings.get_sp_cert
``` 

I got OpenSSL::X509::CertificateError Exception: nested asn1 error  so the format function is not working as expected.

**Pull request merged**